src/verifier/step4/SumcheckLogic.sol

```diff
@@ -0,0 +1,488 @@
+        }
+
+        return MLPoly(P.numvars - 1, new_Z);
+    }
+
+    function evaluate(MLPoly calldata P, uint256[] calldata r) public pure returns (uint256 result) {
+        require(P.Z.length == r.length);
+
+        uint256[] memory chis = EqPolinomialLib.evalsPallas(r);
+
+        for (uint256 i = 0; i < r.length; i++) {
+            // result += chis[i] * P.Z[i];
+            result = addmod(result, mulmod(chis[i], P.Z[i], Pallas.P_MOD), Pallas.P_MOD);
+        }
+    }
+
+    function computeChi(bool[] memory a, uint256[] calldata r) public pure returns (uint256) {
+        require(a.length == r.length);
+
+        uint256 result;
+
+        for (uint256 j = 0; j < r.length; j++) {
+            if (a[j]) {
+                // result *= r[j];
+                result = mulmod(result, r[j], Pallas.P_MOD);
+            } else {
+                // result *= (1 - r[j]);
+                result = mulmod(result, addmod(1, Pallas.negateBase(r[j]), Pallas.P_MOD), Pallas.P_MOD);
+            }
+        }
+
+        return result;
+    }
+
+    function getBits(uint256 num, uint256 numbits) private pure returns (bool[] memory) {
+        bool[] memory result;
+
+        for (uint256 shift_amount = 0; shift_amount < numbits; shift_amount++) {
+            result[shift_amount] = (num & (1 << (numbits - shift_amount - 1)) > 0);
+        }
+
+        return result;
+    }
+
+    function evaluate(SparsePoly calldata P, uint256[] calldata r) public pure returns (uint256) {
+        require(P.numvars == r.length);
+
+        uint256 result;
+
+        for (uint256 i = 0; i < r.length; i++) {
+            bool[] memory bits = getBits(P.Z[i].entry, r.length);
+            // result += computeChi(bits, r) * P.Z[i].idx;
+            result = addmod(result, mulmod(computeChi(bits, r), P.Z[i].idx, Pallas.P_MOD), Pallas.P_MOD);
+        }
+
+        return result;
+    }
+
+    function degree(UniPoly memory poly) public pure returns (uint256) {
+        return poly.coeffs.length - 1;
+    }
+
+    function evalAtZero(UniPoly memory poly) public pure returns (uint256) {
+        return poly.coeffs[0];
+    }
+
+    function evalAtOne(UniPoly memory poly) public pure returns (uint256 result) {
+        for (uint256 i = 0; i < poly.coeffs.length; i++) {
+            // result += poly.coeffs[i];
+            result = addmod(result, poly.coeffs[i], Pallas.R_MOD);
+        }
+    }
+
+    function evaluate(UniPoly memory poly, uint256 r) public pure returns (uint256) {
+        uint256 power = r;
+        uint256 result = poly.coeffs[0];
+        for (uint256 i = 1; i < poly.coeffs.length; i++) {
+            // result += power * poly.coeffs[i];
+            result = addmod(result, mulmod(power, poly.coeffs[i], Pallas.R_MOD), Pallas.R_MOD);
+            // power *= r;
+            power = mulmod(power, r, Pallas.R_MOD);
+        }
+
+        return result;
+    }
+
+    function compress(UniPoly memory poly) public pure returns (CompressedUniPoly memory result) {
```

Comment:
Do we need compression somewhere? I didn't find its usage. If it is still necessary, we need to test it either explicitly by introducing unit-test or implicitly by using it in some higher level function that is tested.